File: scripts/build-share-pages.js
#!/usr/bin/env node
// Generates a real, crawlable URL for every presentation and for the
// About/Sponsor pages: p/<slug>/, about/, sponsor/. GitHub Pages is plain
// static hosting with no server-side logic, so a single index.html can't
// show a different <meta og:image> per shared link - social crawlers read
// whatever's baked into the HTML they fetch and never run JavaScript. So
// each of these is its own small static page with real meta tags baked in
// at generation time, that sends a real visitor into the interactive app
// (index.html, which opens the right drawer/modal from the #presentation/
// <slug>, #about or #sponsor hash) via a redirect crawlers never follow.
//
// Re-run this whenever content.json changes, then redeploy.
'use strict';
const fs = require('fs');
const path = require('path');
const { escapeHtml, stripTags, truncate } = require('./lib/text');

const REPO = path.dirname(__dirname);
const CONTENT = process.argv[2] || path.join(REPO, 'content.json');
const SITE_URL = 'https://fest.philosophers.group';

function page({ title, heading, description, url, ogType, image, imageW, imageH, redirect }) {
  const t = escapeHtml(title);
  const h = escapeHtml(heading);
  const d = escapeHtml(description);
  return `<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>${t}</title>
<meta name="description" content="${d}">
<link rel="canonical" href="${escapeHtml(url)}">
<meta property="og:type" content="${ogType}">
<meta property="og:site_name" content="New Orleans Arts &amp; Ideas Festival">
<meta property="og:title" content="${t}">
<meta property="og:description" content="${d}">
<meta property="og:url" content="${escapeHtml(url)}">
<meta property="og:image" content="${escapeHtml(image)}">
<meta property="og:image:width" content="${imageW}">
<meta property="og:image:height" content="${imageH}">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:title" content="${t}">
<meta name="twitter:description" content="${d}">
<meta name="twitter:image" content="${escapeHtml(image)}">
<meta http-equiv="refresh" content="0; url=${escapeHtml(redirect)}">
<script>location.replace(${JSON.stringify(redirect)});</script>
<style>
  body { font-family: Georgia, serif; background: #f6f3ee; color: #241d15; max-width: 640px; margin: 60px auto; padding: 0 20px; line-height: 1.6; }
  a { color: #a8752c; }
</style>
</head>
<body>
<p>Redirecting to the New Orleans Arts &amp; Ideas Festival schedule&hellip;</p>
<h1>${h}</h1>
<p>${d}</p>
<p><a href="${escapeHtml(redirect)}">Continue to the schedule &rarr;</a></p>
</body>
</html>
`;
}

function writePage(filePath, content) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  fs.writeFileSync(filePath, content);
}

function main() {
  const data = JSON.parse(fs.readFileSync(CONTENT, 'utf8'));
  const mediaById = new Map(data.media.map((m) => [m.id, m]));
  const oggPool = (data.site_images && data.site_images.ogg) || [];

  function imageFor(featuredMediaId, fallbackKey) {
    const m = mediaById.get(featuredMediaId);
    if (m) {
      return [`${SITE_URL}/${m.source_url}`, m.media_details.width, m.media_details.height];
    }
    if (oggPool.length) {
      const pick = oggPool[fallbackKey % oggPool.length];
      return [`${SITE_URL}/${pick}`, 1424, 752];
    }
    return [`${SITE_URL}/media/ogg-arts.png`, 1424, 752];
  }

  let count = 0;
  for (const p of data.presentations) {
    if (!p.slug) continue;
    const title = stripTags(p.title.rendered);
    const excerpt = stripTags((p.excerpt || {}).rendered || '');
    const sf = p.scraped_fields || {};
    const description = truncate(excerpt || sf.presenter_bio || stripTags(p.content.rendered) || title);
    const [image, imageW, imageH] = imageFor(p.featured_media, p.id);
    const url = `${SITE_URL}/p/${p.slug}/`;
    writePage(
      path.join(REPO, 'p', p.slug, 'index.html'),
      page({
        title: `${title} — NOAI`,
        heading: title,
        description,
        url,
        ogType: 'article',
        image, imageW, imageH,
        redirect: `/#presentation/${p.slug}`,
      }),
    );
    count += 1;
  }
  console.log(`wrote ${count} presentation share pages`);

  const pagesById = new Map(data.pages.map((pg) => [pg.id, pg]));
  const pageSpecs = [
    [469, 'about', '#about', 'website'],
    [620, 'sponsor', '#sponsor', 'website'],
  ];
  for (const [pageId, slug, hashTarget, ogType] of pageSpecs) {
    const pg = pagesById.get(pageId);
    if (!pg) continue;
    const title = stripTags(pg.title.rendered);
    const description = truncate(stripTags(pg.content.rendered) || title);
    let [image, imageW, imageH] = imageFor(pg.featured_media, pageId);
    // Prefer the page's own first inline image (e.g. About's gallery) over the random pool.
    const firstImgMatch = /src="(media\/[^"]+)"/.exec(pg.content.rendered || '');
    if (firstImgMatch) {
      const match = data.media.find((m) => m.source_url === firstImgMatch[1]);
      if (match) {
        image = `${SITE_URL}/${firstImgMatch[1]}`;
        imageW = match.media_details.width;
        imageH = match.media_details.height;
      }
    }
    const url = `${SITE_URL}/${slug}/`;
    writePage(
      path.join(REPO, slug, 'index.html'),
      page({
        title: `${title} — NOAI`,
        heading: title,
        description,
        url,
        ogType,
        image, imageW, imageH,
        redirect: `/${hashTarget}`,
      }),
    );
    console.log(`wrote ${slug}/index.html`);
  }
}

main();

File: scripts/build_share_pages.py
#!/usr/bin/env python3
"""Generates a real, crawlable, directly-loadable URL for every presentation
and for the About/Sponsor pages: p/<slug>/, about/, sponsor/.

Each one is a full copy of index.html - the same app, same content.json,
same everything - with only the <head> tags (title, description, canonical,
og:*, twitter:*) swapped for that item's own. That's what makes a shared
link's preview show the right title/description/image, and why it's real
for SEO: a fresh load or a crawler hitting /p/some-slug/ gets that content
immediately, with no client-side redirect. index.html's own <base href="/">
is what lets the exact same relative content.json/media/* references work
correctly no matter which of these directories the file's served from, and
index.html's own JS (openFromLocation) is what notices, from location.
pathname, which presentation/page to open on load.

Re-run this whenever content.json OR index.html changes, then redeploy.
"""
import html
import json
import os
import re
import sys

REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CONTENT = sys.argv[1] if len(sys.argv) > 1 else os.path.join(REPO, 'content.json')
INDEX_HTML = os.path.join(REPO, 'index.html')
SITE_URL = "https://fest.philosophers.group"

TAG_RE = re.compile(r'<[^>]+>')


def clean_text(s):
    return html.unescape(TAG_RE.sub(' ', s or '')).strip()


def truncate(s, n=160):
    s = re.sub(r'\s+', ' ', s).strip()
    return s if len(s) <= n else s[:n - 1].rstrip() + "…"


def set_meta(doc, attr, name, value):
    pattern = re.compile(r'(<meta ' + attr + '="' + re.escape(name) + r'" content=")[^"]*(")')
    new_doc, n = pattern.subn(lambda m: m.group(1) + html.escape(value, quote=True) + m.group(2), doc)
    if n != 1:
        raise ValueError(f"expected exactly one <meta {attr}=\"{name}\"> in index.html, found {n}")
    return new_doc


def set_title(doc, value):
    new_doc, n = re.subn(r'<title>[^<]*</title>', '<title>' + html.escape(value) + '</title>', doc)
    if n != 1:
        raise ValueError(f"expected exactly one <title> in index.html, found {n}")
    return new_doc


def set_canonical(doc, url):
    new_doc, n = re.subn(
        r'(<link rel="canonical" href=")[^"]*(")',
        lambda m: m.group(1) + html.escape(url, quote=True) + m.group(2),
        doc,
    )
    if n != 1:
        raise ValueError(f"expected exactly one <link rel=\"canonical\"> in index.html, found {n}")
    return new_doc


def build_page(template, *, title, description, url, og_type, image, image_w, image_h):
    doc = template
    doc = set_title(doc, title)
    doc = set_meta(doc, 'name', 'description', description)
    doc = set_canonical(doc, url)
    doc = set_meta(doc, 'property', 'og:type', og_type)
    doc = set_meta(doc, 'property', 'og:title', title)
    doc = set_meta(doc, 'property', 'og:description', description)
    doc = set_meta(doc, 'property', 'og:url', url)
    doc = set_meta(doc, 'property', 'og:image', image)
    doc = set_meta(doc, 'property', 'og:image:width', str(image_w))
    doc = set_meta(doc, 'property', 'og:image:height', str(image_h))
    doc = set_meta(doc, 'name', 'twitter:title', title)
    doc = set_meta(doc, 'name', 'twitter:description', description)
    doc = set_meta(doc, 'name', 'twitter:image', image)
    return doc


def write_page(path, content):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, 'w') as f:
        f.write(content)


def main():
    with open(CONTENT) as f:
        data = json.load(f)
    with open(INDEX_HTML) as f:
        template = f.read()

    media_by_id = {m['id']: m for m in data['media']}
    ogg_pool = (data.get('site_images') or {}).get('ogg') or []

    def image_for(featured_media_id, fallback_key):
        m = media_by_id.get(featured_media_id)
        if m:
            return (
                f"{SITE_URL}/{m['source_url']}",
                m['media_details']['width'],
                m['media_details']['height'],
            )
        if ogg_pool:
            pick = ogg_pool[fallback_key % len(ogg_pool)]
            return (f"{SITE_URL}/{pick}", 1424, 752)
        return (f"{SITE_URL}/media/ogg-arts.png", 1424, 752)

    count = 0
    for p in data['presentations']:
        if not p.get('slug'):
            continue
        title = clean_text(p['title']['rendered'])
        sf = p.get('scraped_fields') or {}
        excerpt = clean_text((p.get('excerpt') or {}).get('rendered', ''))
        description = truncate(excerpt or sf.get('presenter_bio') or clean_text(p['content']['rendered']) or title)
        image, iw, ih = image_for(p.get('featured_media'), p['id'])
        url = f"{SITE_URL}/p/{p['slug']}/"
        page = build_page(
            template,
            title=f"{title} — NOAI",
            description=description,
            url=url,
            og_type="article",
            image=image, image_w=iw, image_h=ih,
        )
        write_page(os.path.join(REPO, 'p', p['slug'], 'index.html'), page)
        count += 1
    print(f"wrote {count} presentation share pages")

    page_specs = [
        (469, 'about', 'website'),
        (620, 'sponsor', 'website'),
    ]
    pages_by_id = {pg['id']: pg for pg in data['pages']}
    for page_id, slug, og_type in page_specs:
        pg = pages_by_id.get(page_id)
        if not pg:
            continue
        title = clean_text(pg['title']['rendered'])
        description = truncate(clean_text(pg['content']['rendered']) or title)
        image, iw, ih = image_for(pg.get('featured_media'), page_id)
        # Prefer the page's own first inline image (e.g. About's gallery) over the random pool.
        first_img = re.search(r'src="(media/[^"]+)"', pg['content']['rendered'] or '')
        if first_img:
            local = first_img.group(1)
            match = next((m for m in data['media'] if m['source_url'] == local), None)
            if match:
                image = f"{SITE_URL}/{local}"
                iw, ih = match['media_details']['width'], match['media_details']['height']
        url = f"{SITE_URL}/{slug}/"
        page = build_page(
            template,
            title=f"{title} — NOAI",
            description=description,
            url=url,
            og_type=og_type,
            image=image, image_w=iw, image_h=ih,
        )
        write_page(os.path.join(REPO, slug, 'index.html'), page)
        print(f"wrote {slug}/index.html")


if __name__ == '__main__':
    main()
